moxiworks/widget-factory
# OT-40 / OT-45 / OT-46 - Add models and migrations for widgets

## Description

This adds the following models with migrations:
- `Widget` - the widget data that is shown in the widget admin form
- `UserWidget` - the widgets that each user has added to their dashboard (ranked by `row_order`)
- `UserSetting` - users' widget-specific settings

I added the widget panel, Hurdlr, and ListTrac widgets to the `seeds.rb`.

I also added the widget panel empty state.

## TODO:
- I have not added any code for actually creating/updating anything yet.  I've just used the rails console so far to add/remove `UserWidget` rows.  Nothing is wired up yet for `UserSettings` either.
- JSON endpoints for widget admin
- Somehow reduce duplicated code across components
- Maybe use [variants](https://viewcomponent.org/guide/templates.html#templates) to have separate template files for each component's inline, expanded, and settings views.
- Permissions

## JIRA Link
https://moxiworks.atlassian.net/browse/OT-40

## Validation Steps
As long as you've run `db:migrate` and `db:seed`, the "My Widgets" panel should now show an empty state until you have created a `UserWidget` record associating the `user_uuid` with a `widget_id`.  The widget "Library" modal should render the Hurdlr and ListTrac widgets with descriptions and demo data.  Once a `UserWidget` record is created, the associated widget should have an "Added to Dashboard" button instead of "Add Widget".


![image](https://user-images.githubusercontent.com/3342530/214094089-db404070-b9d2-422e-891e-d5e2049dc967.png)


![image](https://user-images.githubusercontent.com/3342530/214093850-db018b43-0308-40b8-9b17-ff60cb570d41.png)
